docs/lab/aws/dkms-ooto-driver.yaml

```
@@ -0,0 +1,46 @@
+apiVersion: kmm.sigs.k8s.io/v1beta1
+kind: Module
+metadata:
+  name: ooto-ci-a
+spec:
+  driverContainer:
+    container:
+      modprobe:
+        moduleName: ooto-ci-a 
+
+      kernelMappings:
+        - literal: 5.4.196-108.356.amzn2.x86_64
+          containerImage: 804517629211.dkr.ecr.eu-west-3.amazonaws.com/ooto-kmod:1.0-dkms
+          build:
+            buildArgs:
+              - name: CI_BUILD_ARG
+                value: some-build-arg
+            pull:
+              insecure: false
+            push:
+              insecure: false
+              #secrets:
+              #- name: build-secret
+            dockerfile: |
+              ARG KERNEL_VERSION=''
+              FROM amazonlinux as builder
+              RUN amazon-linux-extras install kernel-5.4 -y
+              RUN yum -y install dkms kernel-headers-5.4.196-108.356.amzn2.x86_64 kernel-5.4.196-108.356.amzn2.x86_64 kernel-devel-5.4.196-108.356.amzn2.x86_64 python3 git make gcc bison flex elfutils-libelf-devel linux-headers openssl-devel xz yum-utils bc && yum clean all
+              WORKDIR /usr/src
+              RUN git clone https://github.com/enriquebelarte/ooto-kmod
+              RUN mv ooto-kmod ooto-kmod-0.0.1
+              RUN dkms add ooto-kmod/0.0.1
+              RUN dkms build ooto-kmod/0.0.1
+              RUN dkms install ooto-kmod/0.0.1
```

```suggestion
              RUN git clone https://github.com/enriquebelarte/ooto-kmod ooto-kmod-0.0.1 && \ 
                       dkms add ooto-kmod/0.0.1 && \
                       dkms build ooto-kmod/0.0.1 && \
                       dkms install ooto-kmod/0.0.1
```

Each `RUN` create a new layer in the container, therefore, logical blocks should be fitted in a single `RUN`.

Also, it needs  to be updated to `KMM` instead of `OOT` and cannot point to your github account.

---

This is always the debate - caching vs storage.

The best practice says (I have read it somewhere but can't find it now) that package installations should be in logical blocks, and be ended with cleaning the cache.

In addition to that, most of the examples in [docker-docs](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/) are formed that way.

Below is a well-formed RUN instruction that demonstrates all the apt-get recommendations.
```
RUN apt-get update && apt-get install -y \
    aufs-tools \
    automake \
    build-essential \
    curl \
    dpkg-sig \
    libcap-dev \
    libsqlite3-dev \
    mercurial \
    reprepro \
    ruby1.9.1 \
    ruby1.9.1-dev \
    s3cmd=1.1.* \
 && rm -rf /var/lib/apt/lists/*
```

Here is yet another explanation - https://stackoverflow.com/questions/39223249/multiple-run-vs-single-chained-run-in-dockerfile-which-is-better

---

There is a section in the Best practices Docker doc you linked referring to layers.

---

Thanks. I have seen it but it was more referring to the fact Docker inner implementation has changed to only create layers for `RUN` `COPY` and `ADD` so I didn't think it is a good enough argument :) But sure it helps.

---

My link was to agree with you :) That's why changes have been made.

---

Thanks.